internal/uvm/vpmem.go

```diff
@@ -2,60 +2,301 @@ package uvm
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"os"
 
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
 	"github.com/Microsoft/hcsshim/internal/guestrequest"
 	"github.com/Microsoft/hcsshim/internal/log"
 	"github.com/Microsoft/hcsshim/internal/requesttype"
 	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
-	"github.com/sirupsen/logrus"
 )
 
 const (
-	lcowVPMEMLayerFmt = "/run/layers/p%d"
+	lcowVPMemLayerMountFmt = "/run/layers/p%d-%d-%d"
+	PageSize               = 0x1000
 )
 
 var (
 	// ErrMaxVPMEMLayerSize is the error returned when the size of `hostPath` is
 	// greater than the max vPMEM layer size set at create time.
 	ErrMaxVPMEMLayerSize = errors.New("layer size is to large for VPMEM max size")
+	// ErrNotEnoughSpace is the error returned when a certain VPMem device cannot allocate
+	// requested memory
+	ErrNotEnoughSpace = errors.New("not enough space")
 )
 
+func pageAlign(t uint64) uint64 {
+	if t%PageSize == 0 {
+		return t
+	}
+	return (t/PageSize + 1) * PageSize
+}
+
+func (vpm *vpmemMapping) RefCount() uint32 {
+	return vpm.refCount
+}
+
+func (vpm *vpmemMapping) AddRef() uint32 {
+	vpm.refCount++
```

Comment:
Should increment/decrement be guarded by a lock?

Reply:
we hold a global UVM level lock afaik. do you see scenarios where it might not be enough?

Reply:
I don't think that's a problem. I think we could add a comment on the interface with the ref based funcs that says something like:

"Functions for accessing and updating the reference count of a device. Caller must be holding the UVM global lock when using these functions."

Reply:
I didn't even realize you were still locking with the device lock `m`.